File: backend/services/seal_client.py
# backend/seal/seal_client.py

from typing import Tuple
from sui import SuiClient
from mysten.seal import SealClient
from mysten.sui_bcs import from_hex

# ------------------------------------------------------
# CONFIGURATION — YOU MUST KEEP THESE VALUES
# ------------------------------------------------------

# Your deployed Seal Policy package ID
SEAL_POLICY_PACKAGE = "0xf0ce36656cd421dce66aef95e972e901f21689c3a7ce6402a3b12d4b17eb7b61"

# Your created AllowList shared object ID
ALLOWLIST_OBJECT_ID = "0x85c9763f1fb62d8ff01f33ab681bebf4d7e5891ae18f52a09bb3f01d50b070e6"

# Verified Testnet Key Servers (2-of-N threshold)
KEY_SERVER_IDS = [
    "0x73d05d62c18d9374e3ea529e8e0ed6161da1a141a94d3f76ae3fe4e99356db75",  # Mysten Key Server 1
    "0xf5d14a81a982144ae441cd7d64b09027f116a468bd36e7eca494f750591623c8",  # Mysten Key Server 2
]

# Sui Testnet node
SUI_RPC = "https://fullnode.testnet.sui.io"

# ------------------------------------------------------
# INITIALIZE CLIENTS
# ------------------------------------------------------

# Sui client for Move policy evaluation
sui_client = SuiClient(SUI_RPC)

# Seal Client – this will handle threshold encryption
seal_client = SealClient(
    suiClient=sui_client,
    serverConfigs=[{"objectId": s, "weight": 1} for s in KEY_SERVER_IDS],
    verifyKeyServers=False,
)

# ------------------------------------------------------
# MAIN FUNCTIONS (USED BY BACKEND)
# ------------------------------------------------------

def seal_encrypt(data: bytes, audit_id: str) -> Tuple[bytes, bytes]:
    """
    Encrypt a dataset, model file, or fairness report using Seal.

    audit_id:
        Unique string such as "audit_001", "report_2025_11_21", etc.
        It becomes the identity used for encryption in Seal.

    Returns:
        encrypted_bytes (bytes) – ready to upload to Walrus
        backup_key (bytes) – symmetric DEM fallback key (store or discard)
    """
    encrypted_obj = seal_client.encrypt({
        "threshold": 2,  # Require 2 key servers to decrypt
        "packageId": from_hex(SEAL_POLICY_PACKAGE),
        "id": audit_id.encode("utf-8"),
        "data": data,
    })

    encrypted_bytes = encrypted_obj["encryptedObject"]
    backup_key = encrypted_obj["key"]

    return encrypted_bytes, backup_key


def prepare_identity(audit_number: int) -> str:
    """
    Generates stable identity strings for each encrypted audit item.
    Example: audit_001, audit_002, audit_003...
    """
    return f"audit_{audit_number:03d}"


# ------------------------------------------------------
# OPTIONAL — STORE BACKUP KEYS (IF NEEDED FOR EMERGENCY)
# ------------------------------------------------------

def store_backup_key(audit_id: str, key: bytes):
    """
    Optional fallback: write the backup key to local storage.
    Not required for Seal, but useful for emergency recovery.
    """
    with open(f"backend/tmp/{audit_id}_backup.key", "wb") as f:
        f.write(key)

File: backend/services/seal_client.js
// backend/services/seal_client.js
// --------------------------------------------------------------------
// FINAL WORKING VERSION — SUI SDK FIXED
// --------------------------------------------------------------------
import crypto from "crypto";

if (!globalThis.crypto) {
  globalThis.crypto = crypto.webcrypto;
}

import fs from "fs";
import { SuiClient } from "@mysten/sui/client";
import { fromHEX } from "@mysten/sui/utils";
import { SealClient } from "@mysten/seal";

// --------------------------------------------------------------------
// CONFIG
// --------------------------------------------------------------------

const SEAL_POLICY_PACKAGE =
  "0xf0ce36656cd421dce66aef95e972e901f21689c3a7ce6402a3b12d4b17eb7b61";

const SEAL_ALLOWLIST_OBJECT_ID =
  "0x85c9763f1fb62d8ff01f33ab681bebf4d7e5891ae18f52a09bb3f01d50b070e6";

const KEY_SERVERS = [
  {
    objectId: "0x73d05d62c18d9374e3ea529e8e0ed6161da1a141a94d3f76ae3fe4e99356db75",
    weight: 1,
  },
  {
    objectId: "0xf5d14a81a982144ae441cd7d64b09027f116a468bd36e7eca494f750591623c8",
    weight: 1,
  },
];

const SUI_RPC = "https://fullnode.testnet.sui.io";

// --------------------------------------------------------------------
// CLIENTS
// --------------------------------------------------------------------

const suiClient = new SuiClient({ url: SUI_RPC });

const sealClient = new SealClient({
  suiClient,
  serverConfigs: KEY_SERVERS,
  verifyKeyServers: false,
});

// --------------------------------------------------------------------
// MAIN FUNCTION — encryptFile()
// --------------------------------------------------------------------

async function encryptFile(inputPath, identity) {
  if (!fs.existsSync(inputPath)) {
    console.error("File not found:", inputPath);
    process.exit(1);
  }

  try {
    const raw = fs.readFileSync(inputPath);

    const encrypted = await sealClient.encrypt({
    threshold: 2,
    packageId: SEAL_POLICY_PACKAGE,          // <── FIXED
    policyObjectId: SEAL_ALLOWLIST_OBJECT_ID,
    id: Buffer.from(identity, "utf8").toString("hex"),
    data: raw,
  });

    const final = {
      encryptedObject: Buffer.from(encrypted.encryptedObject).toString("base64"),
      backupKey: Buffer.from(encrypted.key).toString("base64"),
      identity,
    };

    console.log(JSON.stringify(final));
  } catch (err) {
    console.error("Seal Encryption FAILED:", err);
    process.exit(1);
  }
}

// --------------------------------------------------------------------
// CLI ENTRYPOINT
// --------------------------------------------------------------------

const input = process.argv[2];
const identity = process.argv[3];

if (!input || !identity) {
  console.error("Usage: node seal_client.js <file_path> <identity>");
  process.exit(1);
}

encryptFile(input, identity);
